mitiq/observable/pauli.py

```diff
@@ -0,0 +1,128 @@
+        self,
+        *,
+        spec: str,
+        coeff: complex = 1.0,
+        support: Optional[Sequence[int]] = None,
+    ) -> None:
+        """Initialize a PauliString.
+
+        Args:
+            spec: String specifier of the PauliString. Should only contain
+                characters 'I', 'X', 'Y', and 'Z'.
+            coeff: Coefficient of the PauliString.
+            support: Qubits the ``spec`` acts on, if provided.
+
+        Examples:
+            >>> PauliString(spec="IXY")  # X(1)*Y(2)
+            >>> PauliString(spec="ZZ", coeff=-0.5)  # -0.5*Z(0)*Z(1)
+            >>> PauliString(spec="XZ", support=(10, 17))  # X(10)*Z(17)
+        """
+        if not set(spec).issubset(set(self._string_to_gate_map.keys())):
+            raise ValueError(
+                f"One or more invalid characters in spec {spec}. Valid "
+                f"characters are 'I', 'X', 'Y', and 'Z', and the spec should "
+                f"not contain any spaces."
+            )
+        if support is not None:
+            if len(support) != len(spec):
+                raise ValueError(
+                    f"The spec has {len(spec)} Pauli's but the support has "
+                    f"{len(support)} qubits. These numbers must be equal."
+                )
+        else:
+            support = range(len(spec))
+
+        self._pauli = cirq.PauliString(
+            coeff,
+            (
+                self._string_to_gate_map[s].on(cirq.LineQubit(i))
+                for (i, s) in zip(support, spec)
+            ),
+        )
+
+    def matrix(self) -> np.ndarray:
+        """Returns the (potentially very large) matrix of the PauliString."""
+        return self._pauli.matrix()
+
+    def measure_in(self, circuit: QPROGRAM) -> QPROGRAM:
+        @atomic_converter
+        def _measure_in(circuit: cirq.Circuit, pauli: cirq.PauliString):
+            # Transform circuit to canonical qubit layout.
+            qubit_map = dict(
+                zip(
+                    sorted(circuit.all_qubits()),
+                    cirq.LineQubit.range(len(circuit.all_qubits())),
+                )
+            )
+            circuit = circuit.transform_qubits(lambda q: qubit_map[q])
+
+            # Measure the Paulis.
+            if not set(pauli).issubset(set(circuit.all_qubits())):
+                raise ValueError(
+                    f"Qubit mismatch. The PauliString {self} acts on qubits "
+                    f"{[q.x for q in pauli.qubits]} but the circuit has qubit "
+                    f"indices {sorted([q.x for q in circuit.all_qubits()])}."
+                )
+            measured = (
+                circuit + pauli.to_z_basis_ops() + cirq.measure(*pauli.qubits)
+            )
+
+            # Transform circuit back to original qubits.
+            reverse_qubit_map = dict(zip(qubit_map.values(), qubit_map.keys()))
+            return measured.transform_qubits(lambda q: reverse_qubit_map[q])
+
+        return _measure_in(circuit, self._pauli)
```

Comment:
This function `measure_in` is nice but I think it measures the Pauli operator without considering the coefficient.  E.g. it acts in the same way when measuring XZ or 7*XZ. 
I am not sure what is the best solution, e.g.:
- we add a note in the docstring of `measure_in` saying that the coefficient is neglected.
- we could force `PauliString` to always have coefficient 1 and store the coefficients in a different place/object.

Reply:
The coefficient shouldn't go with the circuit but rather the results from executing the circuit. In other words, the circuits you would run to measure XZ or 7*XZ are the same, so they should be the same here. Perhaps there is a better name than `measure_in`? Or I can add a note to the docstring.

Reply:
Verbose, but what this method really does is `add_single_qubit_basis_rotations_to_measure_in(circuit: QPROGRAM) -> QPROGRAM`. Or we could steal the Cirq name of `to_z_basis_ops`. Thoughts?

Reply:
With `to_z_basis_ops` one would expect only the final rotation, not the full circuit. So `measure_in` is fine.